### PR TITLE
Log error when podman client cannot be initialized

### LINKED ISSUE #6501: odo experimental feature --run-on podman fails with wrong error saying 'unable to access podman'
/kind bug

## What versions of software are you using?

**Operating System: MacOS**

**Output of `odo version`: v3.4.0 (b8662ef74)**

## How did you run odo exactly?

* Installed podman using brew
* Added podman to PATH
* odo dev --run-on podman

## Actual behavior

odo reports error as ` ✗  unable to access podman. Do you have podman client installed?` see log below

```
nodejs-ex % ODO_EXPERIMENTAL_MODE=true ~/Sources/vscode-openshift-tools/out/tools/darwin/odo-darwin-amd64 dev --run-on=podman -v 10
I0117 12:31:22.027391   15601 implem.go:109] The path for preference file is /Users/eskimo/.odo/preference.yaml
I0117 12:31:22.027483   15601 util.go:733] HTTPGetRequest: https://raw.githubusercontent.com/redhat-developer/odo/main/build/VERSION
I0117 12:31:22.027523   15601 segment.go:265] Checking telemetry enable status
I0117 12:31:22.027528   15601 segment.go:295] Sending telemetry disabled by preference
I0117 12:31:22.027652   15601 util.go:754] Response will be cached in /var/folders/yz/50s02b8n4yj3_hshg5g6xpz80000gn/T/odohttpcache for 1h0m0s
I0117 12:31:22.028173   15601 util.go:767] Cached response used.
I0117 12:31:22.029192   15601 version.go:28] executing [podman version --format json]
I0117 12:31:22.657304   15601 implem.go:109] The path for preference file is /Users/eskimo/.odo/preference.yaml
Experimental mode enabled for flag: --run-on. Use at your own risk. More details on https://odo.dev/docs/user-guides/advanced/experimental-mode
I0117 12:31:22.660029   15601 context.go:141] absolute devfile path: '/Users/eskimo/Temp/Test1/nodejs-ex/devfile.yaml'
I0117 12:31:22.660055   15601 context.go:92] absolute devfile path: '/Users/eskimo/Temp/Test1/nodejs-ex/devfile.yaml'
I0117 12:31:22.660880   15601 content.go:33] converted devfile YAML to JSON
I0117 12:31:22.661011   15601 apiVersion.go:45] devfile schemaVersion: '2.1.0'
I0117 12:31:22.661042   15601 helper.go:41] devfile apiVersion '2.1.0' is supported
I0117 12:31:22.672108   15601 schema.go:46] validated devfile schema
I0117 12:31:22.675316   15601 validate.go:45] Successfully validated devfile sections
I0117 12:31:22.675455   15601 validate.go:45] Successfully validated devfile sections
I0117 12:31:22.675551   15601 component.go:88] name of component is "nodejs-ex-comp", and sanitized name is "nodejs-ex-comp"
 ✗  unable to access podman. Do you have podman client installed?
```

## Expected behavior

In 3.3.0 release it was reporting correct error, which helped a lot for the first run after podman installed.

```
nodejs-ex % ODO_EXPERIMENTAL_MODE=true ~/Downloads/odo-330 dev --run-on=podman -v 10                                              
Experimental mode enabled for flag: --run-on. Use at your own risk. More details on https://odo.dev/docs/user-guides/advanced/experimental-mode
I0117 12:12:46.082094   15291 implem.go:109] The path for preference file is /Users/eskimo/.odo/preference.yaml
I0117 12:12:46.082324   15291 util.go:733] HTTPGetRequest: https://raw.githubusercontent.com/redhat-developer/odo/main/build/VERSION
I0117 12:12:46.082361   15291 segment.go:265] Checking telemetry enable status
I0117 12:12:46.082368   15291 segment.go:295] Sending telemetry disabled by preference
I0117 12:12:46.082500   15291 util.go:754] Response will be cached in /var/folders/yz/50s02b8n4yj3_hshg5g6xpz80000gn/T/odohttpcache for 1h0m0s
I0117 12:12:46.083378   15291 util.go:767] Cached response used.
I0117 12:12:46.086813   15291 implem.go:109] The path for preference file is /Users/eskimo/.odo/preference.yaml
I0117 12:12:46.087398   15291 context.go:141] absolute devfile path: '/Users/eskimo/Temp/Test1/nodejs-ex/devfile.yaml'
I0117 12:12:46.087430   15291 context.go:92] absolute devfile path: '/Users/eskimo/Temp/Test1/nodejs-ex/devfile.yaml'
I0117 12:12:46.087619   15291 content.go:33] converted devfile YAML to JSON
I0117 12:12:46.087689   15291 apiVersion.go:45] devfile schemaVersion: '2.1.0'
I0117 12:12:46.087758   15291 helper.go:41] devfile apiVersion '2.1.0' is supported
I0117 12:12:46.091748   15291 schema.go:46] validated devfile schema
I0117 12:12:46.092947   15291 validate.go:45] Successfully validated devfile sections
I0117 12:12:46.092997   15291 validate.go:45] Successfully validated devfile sections
I0117 12:12:46.093077   15291 component.go:86] name of component is "nodejs-ex-comp", and sanitized name is "nodejs-ex-comp"
  __
 /  \__     Developing using the "nodejs-ex-comp" Devfile
 \__/  \    Platform: podman
 /  \__/    odo version: v3.3.0
 \__/

↪ Deploying to podman in developer mode
I0117 12:12:46.093357   15291 utils.go:73] Updating container runtime with mandatory volume mounts
 ✗  Deploying pod [626ms]
 ✗  exit status 125: Cannot connect to Podman. Please verify your connection to the Linux system using `podman system connection list`, or try `podman machine init` and `podman machine start` to manage a new Linux VM
Error: unable to connect to Podman socket: Get "http://d/v4.3.1/libpod/_ping": dial unix ///var/folders/yz/50s02b8n4yj3_hshg5g6xpz80000gn/T/podman-run--1/podman/podman.sock: connect: no such file or directory
```

## Any logs, error output, etc?

See above.

### PR DESCRIPTION
**Pull request by valaparthvi**
Signed-off-by: Parthvi Vala <pvala@redhat.com>

**What type of PR is this:**
/kind bug

**What does this PR do / why we need it:**

**Which issue(s) this PR fixes:**

Fixes #6501 

**PR acceptance criteria:**

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**
1. `export PODMAN_CMD=echo`
2. `odo dev --platform podman -v 2`

### CODE REVIEW

File: pkg/podman/errors.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+type PodmanNotFoundError struct {
+	err error
+}
+
+func NewPodmanNotFoundError(err error) PodmanNotFoundError {
+	return PodmanNotFoundError{err: err}
+}
+
+func (o PodmanNotFoundError) Error() string {
+	return fmt.Errorf("unable to access podman. Do you have podman client installed? Cause: %w", o.err).Error()
```
Comment (rm3l) [inline pkg/podman/errors.go:16]: Can `o.err` be nil? In this case, the string returned here might need to be changed.

For example, when running `odo describe component` (which should describe the component on both Podman and the cluster), I got the following message:
```
$ ODO_EXPERIMENTAL_MODE=t PODMAN_CMD=404 odo describe component
============================================================================
⚠ Experimental mode enabled. Use at your own risk.
More details on https://odo.dev/docs/user-guides/advanced/experimental-mode
============================================================================

 ⚠  unable to access podman. Do you have podman client installed? Cause: %!w(<nil>)
 ⚠  failed to get ingresses/routes: ingresses.networking.k8s.io is forbidden: User "system:anonymous" cannot list resource "ingresses" in API group "networking.k8s.io" in the namespace "rh-ee-asoro-dev"
Name: bug-odo-logs-no-cluster
Display Name: Go Runtime
Project Type: Go
Language: Go
Version: 1.0.2
Description: Go is an open source programming language that makes it easy to build simple, reliable, and efficient software.
Tags: Go

Running in: None

Supported odo features:
 •  Dev: true
 •  Deploy: false
 •  Debug: false

Container components:
 •  runtime
    Source Mapping: /projects
```

**Review by rm3l — CHANGES_REQUESTED** [review]
Another point: I find the error message (`executable "echo" not found`) to be misleading for the end-user and might not be accurate. The `echo` executable does exist in my `PATH`, but the output of the command is not the one expected by our Podman Client.
```
$ ODO_EXPERIMENTAL_MODE=t PODMAN_CMD=echo odo describe component --platform podman -v 3     
I0131 17:30:43.309683 3703568 version.go:28] executing [echo version --format json]
 ✗  unable to access podman. Do you have podman client installed? Cause: executable "echo" not found
```

For example, if I try to reproduce the steps provided in #6501 using a real Podman  binary in my `PATH`:
```
$ which podman-v3     
/home/asoro/.local/bin/podman-v3

$ ODO_EXPERIMENTAL_MODE=t PODMAN_CMD=podman-v3 odo describe component --platform podman -v 3
I0131 17:31:21.622000 3705354 version.go:28] executing [podman-v3 version --format json]
 ✗  unable to access podman. Do you have podman client installed? Cause: executable "podman-v3" not found
```
But if I run the same Podman command to retrieve the version:
```
$ podman-v3 version --format json
Cannot connect to Podman. Please verify your connection to the Linux system using `podman system connection list`, or try `podman machine init` and `podman machine start` to manage a new Linux VM
Error: unable to connect to Podman socket: Get "http://d/v3.4.4/libpod/_ping": dial unix ///run/user/1000/podman/podman.sock: connect: no such file or directory
```

I think `odo` should display the full command output if there is an error with the command executed. What do you think?

File: pkg/podman/errors.go (hunk 1 above)
Comment (valaparthvi) [reply pkg/podman/errors.go:16]: > 

I think you raised good points, but I think they might be out of scope for this PR. The errors "executable %q not found" is raised when you initiliaze a new Podman Client. This PR is more about logging error when there is one. I'd be happy to work on it in a separate PR.

**Review by rm3l — COMMENTED** [review]
> > 
> 
> I think you raised good points, but I think they might be out of scope for this PR. The errors "executable %q not found" is raised when you initiliaze a new Podman Client. This PR is more about logging error when there is one. I'd be happy to work on it in a separate PR.

Okay, we can do that in a separate PR. But in this case, I'd keep the original issue (https://github.com/redhat-developer/odo/issues/6501) open, as it was more about `odo` reporting a wrong (and unhelpful) error.
We can say that this PR relates to https://github.com/redhat-developer/odo/issues/6501, but does not close it yet.

**Comment by valaparthvi** [comment]
/override Kubernetes-Integration-Tests/Kubernetes-Docs-Integration-Tests
/override Kubernetes-Integration-Tests/Kubernetes-Integration-Tests
Unrelated errors.

**Comment by valaparthvi** [comment]
> > > 
> > 
> > 
> > I think you raised good points, but I think they might be out of scope for this PR. The errors "executable %q not found" is raised when you initiliaze a new Podman Client. This PR is more about logging error when there is one. I'd be happy to work on it in a separate PR.
> 
> Okay, we can do that in a separate PR. But in this case, I'd leave the original issue (#6501) open, as it was more about `odo` reporting a wrong (and unhelpful) error. We can say that this PR relates to #6501, but, to me, it does not close it yet.

I've updated the NewPodmanCli to return the actual error. But, this will still not work for `odo describe component` where it logs warning  because it simply checks if the podmanClient is initialized.

File: pkg/podman/podman.go [hunk 2]
```diff
@@ -27,7 +27,7 @@ func NewPodmanCli(ctx context.Context) (*PodmanCli, error) {
 	}
 	version, err := cli.Version()
 	if err != nil {
-		return nil, fmt.Errorf("executable %q not found", cli.podmanCmd)
+		return nil, fmt.Errorf("executable %q not found; cause: %w", cli.podmanCmd, err)
```
Comment (rm3l) [inline pkg/podman/podman.go:30]: The `executable %q not found` message might be misleading, as the error is not necessarily due to that.
In this case, I'd suggest returning `err` as it is.

```suggestion
		return nil, err
```

File: pkg/podman/errors.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+type PodmanNotFoundError struct {
+	err error
+}
+
+func NewPodmanNotFoundError(err error) PodmanNotFoundError {
+	return PodmanNotFoundError{err: err}
+}
+
+func (o PodmanNotFoundError) Error() string {
+	msg := "unable to access podman. Do you have podman client installed?"
```
Comment (rm3l) [inline pkg/podman/errors.go:16]: Another suggestion to update this message, which could again be misleading, IMO.
```suggestion
	msg := "unable to access podman. Do you have podman client installed and configured correctly?"
```

**Review by rm3l — COMMENTED** [review]
Added a few comments.

**Review by rm3l — APPROVED** [review]
Thanks Parthvi!

**Pull request merged**